user_tools/src/spark_rapids_dataproc_tools/validation_scripts/metadata_validation.py

```diff
@@ -0,0 +1,196 @@
+
+import time
+import fnmatch
+from pyspark.sql.types import DoubleType
+
+def validation(spark, args):
+
+    result = top_level_metadata(spark, args.format, args.t1, args.t2, args.t1p, args.t2p, args.f)
+    print('-------top level metadata info-------')
+    print(result.show())
+
+    # A result table with the same PK but different values for that column(s)
+    result = metrics_metadata(spark, args.format, args.t1, args.t2, args.t1p, args.t2p, args.pk, args.i, args.e, args.f, args.p)
+    print('-------metadata diff info-------')
+    print(result.show())
+
+    start_time = time.time()
+    print('------------run validation success-----')
+
+def top_level_metadata(spark, format, t1, t2, t1p, t2p, f):
+    if format in ['parquet', 'orc', 'csv']:
+        print('todo')
+    elif format == "hive":
+        results = []
+        table_names = [t1, t2]
+        where_clause = ''
+        if t1p != 'None' and f != 'None':
+            where_clause = f" where {t1p} and {f}"
+        elif t1p != 'None':
+            where_clause = f" where {t1p}"
+        elif f != 'None':
+            where_clause = f" where {f}"
+        for table_name in table_names:
+            sql = f'select * from {table_name}'
+            sql += where_clause
+            df = spark.sql(sql)
+            row_count = df.count()
+            col_count = len(df.columns)
+            results.append((table_name, row_count, col_count))
+        resultsDF = spark.createDataFrame(results, ["TableName", "RowCount", "ColumnCount"])
+        return resultsDF
+
+def generate_metric_df(spark, table_DF, i, t1):
+    result = None
+    agg_functions = [min, max, avg, stddev, countDistinct]
+    # if not specified any included_columns, then get all numeric cols
+    metrics_cols = i
+    if i in ['None', 'all']:
+        metrics_cols = [c.name for c in table_DF.schema.fields if
+                        any(fnmatch.fnmatch(c.dataType.simpleString(), pattern) for pattern in ['*int*', '*decimal*', '*float*', '*double*'])]
+    for col in metrics_cols:
+        dfc = spark.createDataFrame(([col],), ["ColumnName"])
+        table1_agg = table_DF.select(
+            [f(col).cast(DoubleType()).alias(f.__name__ + t1) for f in
+             agg_functions])
+        tmp_df = dfc.join(table1_agg)
+        if result is None:
+            result = tmp_df
+        else:
+            result = result.union(tmp_df)
+    return result
+
+def metrics_metadata(spark, format, t1, t2, t1p, t2p, pk, i, e, f, p):
+    # spark, format, t1, t1p, pk, e, i, f, view_name
+    table1_DF = load_table(spark, format, t1, t1p, pk, e, i, f, "")
+    table2_DF = load_table(spark, format, t2, t2p, pk, e, i, f, "")
+
+    table_metric_df1 = generate_metric_df(spark, table1_DF, i, t1)
+    table_metric_df2 = generate_metric_df(spark, table2_DF, i, t2)
+    joined_table = table_metric_df1.alias("t1").join(table_metric_df2.alias("t2"), ["ColumnName"])
+
+    cond = (round("t1.min" + t1, p) != round("t2.min" + t2, 4)) | \
```

Comment:
I think precision for t2 needs to be changed.

Reply:
updated